### PR TITLE
Bugfix/519 476 user directory usage

### LINKED ISSUE #519: Python Algorithms on windows don't have DCSTOOL_HOME environment
**Describe the bug**
When trying to run python algorithms on windows the PyAlgoEnv.py file on line ten uses os.environ instead of java->System.getProperty to retreieve DCSTOOL_HOME

If this happens a "None Type cannot be added to str" happens, which is confusing.

**To Reproduce**
1. Fresh install,  DCSTOOL_HOME user or system environment does not exist.
2. Create or otherwise run a python algorithm.
3. See error

**Expected behavior**
System should just work. Either the installer should create the variable, or like the other components of opendcs System.getProperty, or preferably `EnvExpander.expand` to get the value.

**Screenshots**

**Desktop/ServerSide (please complete the following information):**
 - OS: Windows
 - Version 7.0.x

**Additional context**

### PR DESCRIPTION
**Pull request by MikeNeilson**
## Problem Description

Fixes #519 and #476. 

Describe the problem you are trying to solve.

## Solution

#519 - Use environment to allow user override, fall back to properties if not set, decj or decj.bat will always set the Java Property

#476 - Scripts now determine there location to find the appropriate decj instance in unix like they do in windows.
Additionally added logic to decj and decj.bat to create an appropriate default instance.

$HOME/.opendcs in unix land
%APPDATA%\opendcs in windows. (This maps to the User's "AppData\Roaming" directory.

I also made the LRGS scripts take a hard stance on LRGSHOME being set and preferably to a location outside the install directory However, if set it doesn't care, the logic only triggers if an LRGSHOME environment variable is not set.

## how you tested the change

Manually exercising the scripts on windows and Linux to very operations.

## Where the following done:

- [ ] Tests. Check all that apply:
   - [ ] Unit tests created or modified that run during ant test.
   - [ ] Integration tests created or modified that run during integration testing
         (Formerly called regression tests.)
   - [ ] Test procedure descriptions for manual testing
- [ ] Was relevant documentation updated?
- [ ] Were relevant config element (e.g. XML data) updated as appropriate

If you aren't sure leave unchecked and we will help guide you to want needs changing where.

### CODE REVIEW

File: install/bin/decj [hunk 1]
```diff
@@ -25,16 +35,10 @@ fi
...
-  elif [ -d $HOME/opendcs_user ]
-  then
-    DCSTOOL_USERDIR=$HOME/opendcs_user
-  else
-    DCSTOOL_USERDIR=$HOME
-  fi
+  DCSTOOL_USERDIR=$HOME/.opendcs
+  echo "Creating Local User Directory and initial properties in ${DCSTOOL_USERDIR}"
+  mkdir -p $DCSTOOL_USERDIR
+  cp $DH/decodes.properties $DCSTOOL_USERDIR/user.properties
```
Comment (ktarbet) [inline install/bin/decj:41]: This should probably be part of install ;  If this runs every time that may clobber data.
May need to update documentation.

Comment (ktarbet) [reply install/bin/decj:41]: OH, I see this only runs if DCSTOOL_USERDIR doesn't exist.

File: install/bin/decj.bat [hunk 2]
```diff
@@ -21,4 +21,18 @@ for /R "%APP_PATH%/dep" %%a in (*.jar) do (
 
 set "CLASSPATH=!CLASSPATH!"
 
-java -Xmx240m -cp "!CLASSPATH!" -DDCSTOOL_HOME="%APP_PATH%" -DDECODES_INSTALL_DIR="%APP_PATH%" -DDCSTOOL_USERDIR="%APP_PATH%" %*%
+if not defined DCSTOOL_USERDIR (
+  set DCSTOOL_USERDIR="%APPDATA%\opendcs"
```
Comment (ktarbet) [inline install/bin/decj.bat:25]: consider making this consistent with Unix,   leading dot    '.opendcs'

File: install/bin/editPasswd [hunk 3]
```diff
@@ -1,2 +1,6 @@
+#!/bin/bash
+SCRIPT=$(readlink -f "$0")
+# Absolute path this script is in, thus /home/user/bin
+APP_PATH=$(dirname "$SCRIPT")
 
-%INSTALL_PATH/bin/decj ilex.util.PasswordFileEditor -f $LRGSHOME/.lrgs.passwd
+${APP_PATH}/decj ilex.util.PasswordFileEditor -f $LRGSHOME/.lrgs.passwd
```
Comment (ktarbet) [inline install/bin/editPasswd:6]: Should we handle checking for $LRGSHOME like done in ./elp below.

File: install/bin/postInstall
Comment (ktarbet) [inline install/bin/postInstall:1]: How about deleting this file?  (and corresponding .bat version)    jars should be managed with dependency tools

File: install/bin/decj (hunk 1 above)
Comment (MikeNeilson) [reply install/bin/decj:41]: Though I do need to check that the folder doesn't exist already. I think I forgot that part.

File: install/bin/editPasswd (hunk 3 above)
Comment (MikeNeilson) [reply install/bin/editPasswd:6]: Yep, good catch.

File: install/bin/decj (hunk 1 above)
Comment (MikeNeilson) [reply install/bin/decj:41]: Also didn't want to put it as part of the install as we may be getting away from the installer.

File: install/bin/postInstall
Comment (MikeNeilson) [reply install/bin/postInstall:1]: Yeah. I'm not even sure it's actually executed anymore. Certainly not as part of the build anyways.

File: install/bin/decj.bat [hunk 4]
```diff
@@ -21,4 +21,18 @@ for /R "%APP_PATH%/dep" %%a in (*.jar) do (
...
 
-java -Xmx240m -cp "!CLASSPATH!" -DDCSTOOL_HOME="%APP_PATH%" -DDECODES_INSTALL_DIR="%APP_PATH%" -DDCSTOOL_USERDIR="%APP_PATH%" %*%
+if not defined DCSTOOL_USERDIR (
+  set DCSTOOL_USERDIR="%APPDATA%\.opendcs"
+)
+
+if not exist %DCSTOOL_USERDIR%\ (
+  echo "Creating Local User Directory and initial properties in %DCSTOOL_USERDIR%"
+  mkdir %DCSTOOL_USERDIR%
+  cp %APP_PATH%\decodes.properties %DCSTOOL_USERDIR%\user.properties
```
Comment (ktarbet) [inline install/bin/decj.bat:31]:
```suggestion
  copy %APP_PATH%\decodes.properties %DCSTOOL_USERDIR%\user.properties
```

File: izpack/opendcs-install.xml
Comment (ktarbet) [inline izpack/opendcs-install.xml:1]: consider removing all lines that have:
```xml
<parsable .... />
```

 the *dynamic* batch and bash scripts don't need to be parsed during install.

File: install/bin/decj.bat (hunk 4 above)
Comment (MikeNeilson) [reply install/bin/decj.bat:31]: Good catch.

File: docs/source/start-software.rst [hunk 5]
```diff
@@ -270,12 +272,31 @@ file into a browser).
...
+
+
+This allows new users to get started more quickly. Additionally if you are setting up a fresh SQL based database, the
+contents of `$DCSTOOL_USERDIR/edit-db` should be imported in, starting with the `loading-app` directory.
+
+.. NOTE::
+   
+   We are currently in the process of improving how this required standard data is handled. We've provided this basic
+   instruction to help users get started but would preferred not to write out detail documentation since we'll be changing it soon.
+   If you have further questions please use github.com/opendcs/opendcs/discourse for more clarity.
```
Comment (ktarbet) [inline docs/source/start-software.rst:293]:
```suggestion
   If you have further questions please use https://github.com/opendcs/opendcs/discussions for more clarity.
```

File: install/bin/decj [hunk 6]
```diff
@@ -5,16 +5,26 @@
...
+DH=$(realpath "${APP_PATH}/..")
 
 # Make sure DCSTOOL_HOME is set
 if [ -z "$DCSTOOL_HOME" ]
 then
   DCSTOOL_HOME=$DH
   export DCSTOOL_HOME
+elif [ "$DH" != "$DCSTOOL_HOME" ]
+then
+  echo "Environment has set a diffent DCSTOOL_HOME than the location of the application start script."
```
Comment (ktarbet) [inline install/bin/decj:21]:
```suggestion
  echo "Environment has set a different DCSTOOL_HOME than the location of the application start script."
```

File: install/bin/elp-local [hunk 7]
```diff
@@ -1,12 +1,14 @@
...
+# Absolute path this script is in, thus /home/user/bin
+APP_PATH=$(dirname "$SCRIPT")
 
-DH=%INSTALL_PATH
 
 if [ -z "$LRGSHOME" ] 
 then
-  LRGSHOME=$DH
-  export LRGSHOME
+  echo "the environment variable 'LRGSHOME' should be set to a path outside the installation directory."
```
Comment (ktarbet) [inline install/bin/elp-local:10]:
```suggestion
  echo "Error: The environment variable 'LRGSHOME' is not set.   Set to a path outside the installation directory."
```

File: install/bin/decj.bat [hunk 8]
```diff
@@ -21,4 +21,20 @@ for /R "%APP_PATH%/dep" %%a in (*.jar) do (
...
+if not defined DCSTOOL_USERDIR (
+  set DCSTOOL_USERDIR="%APPDATA%\.opendcs"
+)
+
+if not exist %DCSTOOL_USERDIR%\ (
+  echo "Creating Local User Directory and initial properties in %DCSTOOL_USERDIR%"
+  echo "The default XML database has been copied to this directory."
+  mkdir %DCSTOOL_USERDIR%
+  copy %APP_PATH%\decodes.properties %DCSTOOL_USERDIR%\user.properties
+  copy /R %APP_PATH%\edit-db %DCSTOOL_USERDIR%\edit-db
```
Comment (ktarbet) [inline install/bin/decj.bat:33]:
```suggestion
  xcopy %APP_PATH%\edit-db %DCSTOOL_USERDIR%\edit-db /E /I
```
unfortunately /R is not part of copy

xcopy /?

/E           Copies directories and subdirectories, including empty ones.
 /I           If destination does not exist and copying more than one file,
               assumes that destination must be a directory.

File: docs/source/dev-docs.rst [hunk 9]
```diff
@@ -127,6 +127,40 @@ The following workflow can be used:
...
+
+    If this is a major concern you should set the DCSTOOL_USERDIR for the session ant runs in
+    to point to a directory that only contains profiles that connect to test systems.
+
+.. code-block:: bash
+    # to just run the launcher
+    ant run
+
+    # to run a specific app
+    ant run -Dopendcs.app=compedit
```
Comment (ktarbet) [inline docs/source/dev-docs.rst:149]: note: the run target seems to assume you have first done a build, such as 'ant opendcs'

**Review by ktarbet — APPROVED** [review]

File: docs/source/dev-docs.rst (hunk 9 above)
Comment (MikeNeilson) [reply docs/source/dev-docs.rst:149]: There's a dependency on "stage" that will force a build to happen. so this is safe to run before `ant jar`.

**Comment by MikeNeilson** [comment]
@ktarbet Since we're creating a default profile with the XML database in the new user directory, should we just go ahead and do the initial baseline/metadata import?

It's not trivial to start setting everything up manually and you have to know what to import. Seems like doing it here would:
1. Simplify the initial setup process more
2. Provide at least some maintained guidance on what to do.

Future work to better organize the metadata setup would just be used here as well so it's not like it's going to be even more to maintain.

**Comment by ktarbet** [comment]
@MikeNeilson 

During the install process, yes, it would be good to install the XML database.  That would be cleaner that having the scripts in ./bin do a just-in-time install.

I think separation of user data location from the application install location is worth the mental model change that will be required for users.  

Users that have an existing XML database may need to manually adapt, and another installed XML database could be confusing.


I'm concerned we may be doing too much in a single PR if we keep adding changes.

**Comment by MikeNeilson** [comment]
I agree with your point about this PR already getting somewhat large.

Though not with during the installer, I think.

You've suggested just a zip, so then when would it happen?
That would be additional work to add to the installer, though it wouldn't be super difficult, basically just the same logic we currently have in the decj files + dbimport.

**Pull request merged**